flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -861,6 +862,25 @@ struct ConvertOpConversion : public FIROpConversion<fir::ConvertOp> {
                                                              ic, one);
       return mlir::success();
     }
+
+    // Follow UNIX F77 convention for logicals:
+    // 1. underlying integer is not zero => logical is .TRUE.
+    // 2. logical is .TRUE. => set underlying integer to 1.
+    auto isI1Type = [](mlir::Type ty) {
+      return ty.isa<mlir::IntegerType>() &&
+             mlir::LLVM::getPrimitiveTypeSizeInBits(ty) == 1;
+    };
```

Comment:
There is exactly one instance of `i1` type in the MLIRContext, so a simple comparison is all that is needed.

```c++
  i1Ty = mlir::IntegerType::get(ctx, 1);
  if (typeOfThing == i1Ty) ...
```